### PR TITLE
fix some errors in github action, and add python 3.12

### PR DESCRIPTION
**Pull request by graphenn**
Not all issues fixed. The left issue can be test with code below.

```
from cyvcf2 import VCF, Writer

VCF_PATH ="./cyvcf2/tests/test.vcf.gz"

def test():
    input_vcf = VCF('./cyvcf2/tests/seg.vcf.gz')
    output_vcf = Writer('./test.vcf', input_vcf)
    for v in input_vcf:
        v.genotypes = [[1,1,False]]
        output_vcf.write_record(v)
    output_vcf.close()

def test2():
    vcf = VCF(VCF_PATH)
    for variant in vcf:
        print(variant)
        break

    print("11")
    vcf.close()
    print("22")

test()
test2()
```

It may be same issue with these, on htslib 1.18

Fixed possible double frees when handling errors in bcf_hdr_add_hrec(), if particular memory allocations fail. (PR https://github.com/samtools/htslib/pull/1637)

Ensure that bcf_hdr_remove() clears up all pointers to the items removed from dictionaries. Failing to do this could have resulted in a call requesting a deleted item via bcf_hdr_get_hrec() returning a stale pointer. (PR https://github.com/samtools/htslib/pull/1637)

I tried but failed to upgrade htslib version.

### CODE REVIEW

File: cyvcf2/cyvcf2.pyx [hunk 1]
```diff
@@ -407,7 +407,7 @@ cdef class VCF(HTSFile):
         itr = bcf_itr_querys(self.hidx, self.hdr, to_bytes(region))
         if itr == NULL:
             warnings.warn("no intervals found for %s at %s" % (self.fname, region))
-            raise StopIteration
+            return
```
Comment (brentp) [inline cyvcf2/cyvcf2.pyx:410]: why are you changing these from StopIteration?

Comment (brentp) [reply cyvcf2/cyvcf2.pyx:410]: thanks very much for explaining.

**Comment by graphenn** [comment]
I have found a solution to decouple the compilation of cyvcf2 and htslib, which now enables the upgrade of htslib. Additionally, by statically linking to the htslib library, we can avoid redundant compilations. This approach ensures a smoother and more efficient development process. I have tested htslib 1.18, the test_write_missing_contig test case still not error.

Do you want me to submit this changes here or later?

**Comment by graphenn** [comment]
everything works now! And I will submit decouple the compilation of cyvcf2 and htslib at another PR later.

**Comment by brentp** [comment]
This is amazing! thanks so much. PR for decoupling would be great!

**Comment by brentp** [comment]
not all heroes wear capes, but, fittingly, your avatar does. :superhero:

**Pull request merged**